log/logger.go

```diff
@@ -42,34 +121,206 @@ func File(path string) (logger *Logger, loggerClose func(), err error) {
 }
 
 // STDOUT returns a logger that writes to STDOUT.
 func STDOUT() (logger *Logger) {
-	return newLoggerWithWriter(os.Stdout)
+	return newLoggerWithWriter(os.Stdout, FlagStandard)
 }
 
-// WithFile returns a logger that writes to a file and to the parent logger at the same time.
-func WithFile(parent *Logger, filePath string) (logger *Logger, loggerClose func(), err error) {
+// newLogWriter returns a logger that writes to a file and to the parent logger at the same time.
+func newLogWriter(parent io.Writer, filePath string) (writer io.Writer, err error) {
 	if err := os.MkdirAll(filepath.Dir(filePath), 0755); err != nil {
-		return nil, nil, pkgerrors.WithStack(err)
+		return nil, pkgerrors.WithStack(err)
 	}
 
 	file, err := os.OpenFile(filePath, os.O_APPEND|os.O_WRONLY|os.O_CREATE, 0644)
 	if err != nil {
-		return nil, nil, pkgerrors.WithStack(err)
+		return nil, pkgerrors.WithStack(err)
 	}
-	loggerClose = func() {
-		if err := file.Close(); err != nil {
-			panic(err)
+
+	writer = io.MultiWriter(parent, file)
+
+	return writer, nil
+}
+
+// spawningHandler is a structural logging handler which spawns a new log file if one of the given log file spawners triggers.
+type spawningHandler struct {
+	// writer holds the writer to write the output.
+	writer io.Writer
+
+	// attributes holds the attributes handed to the logger.
+	attributes map[AttributeKey]string
+
+	// logFileSpawners holds the spawners responsible for spawning a new log file.
+	logFileSpawners []logFileSpawner
+
+	// flags holds the flags deciding which fields are logged.
+	flags Flags
+}
+
+// newSpawningHandler returns a new spawning handler.
+func newSpawningHandler(writer io.Writer, flags Flags) *spawningHandler {
+	return &spawningHandler{
+		writer: writer,
+
+		attributes: map[AttributeKey]string{},
+
+		flags: flags,
+	}
+}
+
+// Clone returns a copy of the object.
+func (h *spawningHandler) Clone() (clone *spawningHandler) {
+	return &spawningHandler{
+		writer: h.writer,
+
+		attributes: maps.Clone(h.attributes),
+
+		logFileSpawners: slices.Clone(h.logFileSpawners),
+	}
+}
+
+var _ slog.Handler = (*spawningHandler)(nil)
+
+// Enabled reports whether the handler handles records at the given level.
+func (h *spawningHandler) Enabled(ctx context.Context, level slog.Level) bool {
+	return true
+}
+
+// Handle handles the Record.
+func (h *spawningHandler) Handle(ctx context.Context, record slog.Record) error {
+	if h.flags&FlagDate != 0 {
+		fmt.Fprint(h.writer, record.Time.Format("2006/01/02"))
+		fmt.Fprint(h.writer, " ")
+	}
+	if h.flags&FlagTime != 0 {
+		fmt.Fprint(h.writer, record.Time.Format("15:04:05"))
+		fmt.Fprint(h.writer, " ")
+	}
+
+	fmt.Fprintln(h.writer, record.Message)
+
+	return nil
+}
+
+// WithAttrs returns a new Handler whose attributes consist of both the receiver's attributes and the arguments.
+// The Handler owns the slice: it may retain, modify or discard it.
+func (h *spawningHandler) WithAttrs(attributes []slog.Attr) slog.Handler {
```

Comment:
So the moment we create a new handler with attributes we check if there needs to be a file that needs to be created... Only thing that is not handled is what happens to the attributes when we don't have enough to start a file... cause those values are then kinda lost...

so maybe we have a slice `unusedAttributes`  that contains the attributes which are not yet required by any spawner... and then those attributes are printed with the logs... though we hopefully won't have this exact case in our setup so we might as well ignore it

Reply:
Yes they are "lost". Right now every log message has still all relevant information in the message itself. E.g 

```
2024/07/08 09:39:00 Evaluated model "symflower/symbolic-execution" on task "write-tests" using language "java" and repository "java/plain": encountered 0 problems: []
```
model, repository, language and task is still contained. but at one point we should switch to something like

```
2024/07/08 09:39:00 [model=symflower/symbolic-execution][language=java][task=write-tests][repository=java/plain] Evaluation done. Encountered 0 problems: []
```

And e.g. the log file for a specific task then contains the same log message but without the "task" attribute printed.